lua/copilot/panel.lua

```diff
@@ -0,0 +1,500 @@
+
+    for i = 1, #lines, 1 do
+      if lines[i]:find(marker_prefix, 0, true) then
+        marker_line = lines[i]
+        marker_linenr = linenr + i
+        break
+      end
+    end
+  else
+    local lines = vim.api.nvim_buf_get_lines(self.bufnr, 0, linenr, false)
+    for i = #lines + dir, 1, -1 do
+      if lines[i]:find(marker_prefix, 0, true) then
+        marker_line = lines[i]
+        marker_linenr = i
+        break
+      end
+    end
+  end
+
+  if not marker_line then
+    return
+  end
+
+  local id = string.match(marker_line, ":id:(.-):", #marker_prefix)
+
+  return self:get_entry(id), marker_linenr
+end
+
+---@param dir -1|0|1
+function panel:jump(dir)
+  local _, linenr = self:get_entry(dir)
+  if linenr then
+    vim.api.nvim_win_set_cursor(self.winid, { linenr, 0 })
+  end
+end
+
+function panel:accept()
+  local entry = self:get_entry(0)
+  if not entry then
+    return
+  end
+
+  local bufnr = vim.uri_to_bufnr(panel_uri_to_doc_uri(self.panel_uri))
+
+  if not vim.api.nvim_buf_is_loaded(bufnr) then
+    vim.cmd("echoerr 'Buffer was closed'")
+    return
+  end
+
+  if vim.fn.getbufline(bufnr, entry.range.start.line + 1)[1] ~= self.state.line then
+    vim.cmd('echoerr "Buffer has changed since synthesizing solution"')
+    return
+  end
+
+  vim.api.nvim_buf_call(bufnr, function()
+    vim.lsp.util.apply_text_edits({
+      { range = entry.range, newText = entry.completionText },
+    }, 0, "utf-16")
+
+    if self.state.was_insert then
+      vim.cmd("startinsert!")
+    else
+      vim.cmd("normal! $")
+    end
+  end)
+
+  self:unlock():clear():lock()
+
+  vim.api.nvim_win_close(self.winid, true)
+end
+
+local function set_keymap(bufnr)
+  if panel.keymap.accept then
+    vim.keymap.set("n", panel.keymap.accept, mod.accept, {
+      buffer = bufnr,
+      desc = "[copilot] (panel) accept",
+      silent = true,
+    })
+  end
+
+  if panel.keymap.jump_prev then
+    vim.keymap.set("n", panel.keymap.jump_prev, mod.jump_prev, {
+      buffer = bufnr,
+      desc = "[copilot] (panel) jump prev",
+      silent = true,
+    })
+  end
+
+  if panel.keymap.jump_next then
+    vim.keymap.set("n", panel.keymap.jump_next, mod.jump_next, {
+      buffer = bufnr,
+      desc = "[copilot] (panel) jump next",
+      silent = true,
+    })
+  end
+
+  if panel.keymap.refresh then
+    vim.keymap.set("n", panel.keymap.refresh, mod.refresh, {
+      buffer = bufnr,
+      desc = "[copilot] (panel) refresh",
+      silent = true,
+    })
+  end
+end
+
+function panel:ensure_bufnr()
+  if self.bufnr and vim.api.nvim_buf_is_valid(self.bufnr) then
+    return
+  end
+
+  self.bufnr = vim.api.nvim_create_buf(false, true)
+
+  for name, value in pairs({
+    bufhidden = "hide",
+    buflisted = false,
+    buftype = "nofile",
+    filetype = "markdown",
+    modifiable = false,
+    readonly = true,
+    swapfile = false,
+    undolevels = 0,
+  }) do
+    vim.api.nvim_buf_set_option(self.bufnr, name, value)
+  end
```

Comment:
Wrapping this in a vim.schedule seemed to lessen the number of highlighting errors I got

Reply:
Can you share repro steps for getting those highlighting errors? I don't think buffer options should be set in `vim.schedule`, they should be set as soon as the buffer is created and before it's being used.

Reply:
> Can you share repro steps for getting those highlighting errors? I don't think buffer options should be set in `vim.schedule`, they should be set as soon as the buffer is created and before it's being used.

All I did was open panel.lua and run `:CopilotPanel`, move the cursor between them a couple times while it was loading for good measure, `:q` and do it again.

You can probably get away with not scheduling the buffer options, but honestly it won’t hurt, and the window ones definitely need to be scheduled based on what occurred when I quit and reopened the window. Even line numbers started appearing…

Reply:
😂 ah this is gonna be tricky. I did that a few times, but still can't reproduce it.

> he window ones definitely need to be scheduled based on what occurred when I quit and reopened the window. Even line numbers started appearing…

Scheduling is only hiding the root issue. There must be something wrong somewhere else. Not solving the root issue would make it appear again in some weird scenarios.

When you do `:q` or close the window in some other way, the `"[copilot] (panel) win closed cleanup"` autocmd is supposed to clean everything up and the next time you open the panel, it will create a brand new window and set it up again. Since I can't reproduce it, I really can't pinpoint the exact reason why it's happening.

---

Are you using the deafult config? Or did you enable `auto_refresh`?

Also just to make sure, are you using the exact commit in this PR? I force pushed many times here, so maybe there's a chance you checked out an older version? 😂

Reply:
I pulled down the changes fresh tonight so I think it's the correct version. I checked that my parsers were up to date too for treesitter so I have no clue 🙃 

It's possible my neovim version is causing problems since I build from source pretty regularly. 

I went with scheduling because when I debugged the the functions that are supposed to be triggered everything seemed to be working perfectly, so I assumed it had to do with cursor location or split(). I've had similar issues with split in the past that were fixed by scheduling the functions meant to run after the split was completed.

Reply:
> It's possible my neovim version is causing problems since I build from source pretty regularly.

Same here. I'm using the latest HEAD, compiled today. 😂

Any chance you can record a short clip when you get the time?

Maybe I'll notice something there? Also seeing the exact error messages would definitely help.

Reply:
> Just thought about it. Can you reproduce the issue with minimal config?
> 
> Maybe you have some autocommands that are overwriting the buf/win options? That might explain why doing `vim.schedule` solves it for you. Without `vim.schedule`, `copilot.panel` is setting buf/win options first, then those autocmds overwriting them. With `vim.schedule`, those autocmds are setting buf/win options first, then `copilot.panel` is overwriting them.
> 
> This is just a hunch, you'll need to verify it by running it with a minimal config.

Hmm, I am pretty careful about that type of stuff, but I can no longer seem to reproduce the problem, so I think that we should just merge this with the changes to conceal the code delimiters. If issues start cropping up, we will have more info to deal with it that way. I think that your code itself is fine, the issue probably was an edge case due to copilot itself (I bet it had markdown quotes in the response somewhere)

Reply:
> I think that your code itself is fine

Just to be clear, I mostly just translated `copilot.vim`'s vimscript code to lua. So most of the stuffs are one-to-one related to the original plugin. I can't take credit for that 😂